Cancel Complete fixes

jkeon opened the pull request:
Cancel Completes used to just return `Entity` instances but that made it difficult to merge into generic jobs that also took in `IEntityProxyInstance`.

To make the API easier to work with, there is a `CancelCompleted` struct that is an `IEntityProxyInstance` that is returned now instead.

### What issues does this resolve?
- None

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [x] Yes - Change all Cancel Complete jobs to take in `CancelCompleted` instead of just `Entity` 
 - [ ] No

mbaker3 commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/DataStream/Cancellation/CancelCompleted.cs:
```diff
@@ -0,0 +1,24 @@
+    /// Represents the completion of a Cancel Request from a generic perspective where only the <see cref="Entity"/>
+    /// is returned that completed it's cancellation flow.
+    /// </summary>
+    public readonly struct CancelCompleted : IEntityProxyInstance
```
Better matches the tense in the rest of the naming in tasks

```suggestion
    public readonly struct CancelComplete : IEntityProxyInstance
```

jkeon replied:
Fixed

mbaker3 commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/DataStream/Cancellation/CancelProgressFlowNode.cs:
```diff
@@ -17,7 +17,7 @@ internal class CancelProgressFlowNode : AbstractAnvilBase
         private readonly CancelProgressFlowNode m_Parent;
         private readonly ActiveLookupData<EntityProxyInstanceID> m_ProgressLookupData;
         private readonly ActiveLookupData<EntityProxyInstanceID> m_ParentProgressLookupData;
```
Maybe a future consideration to make `EntityProxyInstanceID` implement `IEntityProxyInstance` or give each bucket its own data type. That would make opening up access a lot easier in the future.

jkeon replied:
Added TODO and linked to #136

mbaker3 commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/JobDataInteraction/DataStreamCancellationUpdater.cs:
```diff
@@ -177,7 +177,7 @@ private void Debug_EnsureCanContinue(ref TInstance instance)
             if (m_State == UpdaterState.Ready)
             {
-                throw new InvalidOperationException($"Attempting to call {nameof(Continue)} on a {instance} but that element didn't come from this {nameof(DataStreamUpdater<TInstance>)}. Please ensure that the indexer was called first.");
+                throw new InvalidOperationException($"Attempting to call {nameof(Continue)} on a {instance} but our state is {m_State}. Most likely {nameof(Continue)} was called twice or {nameof(Resolve)} was called already for this instance. Alternatively, this instance may not belong to this {nameof(DataStreamCancellationUpdater<TInstance>)}");
```
Better wording?

``` csharp
$"Attempting to call {nameof(Continue)} on a {instance} but the state is {m_State}. Most likely {nameof(Continue)} or {nameof(Resolve)} were called twice for this instance. Alternatively, this instance may not belong to this {nameof(DataStreamCancellationUpdater<TInstance>)}"
```

jkeon replied:
Hmmn in this specific instance there are three reasons this could have happened:

1. You called Continue twice in a row.
2. You called Resolve once and then Continue.
3. This instance isn't part of it. 

If you called Resolve twice then you'd hit the other error message which just reverses Continue/Resolve.

mbaker3 replied:
ah ok. my bad.

mbaker3 reviewed (changes requested):
Some 🚲  comments but the main change required is to allow consumers to trigger an `ITaskJobForDefer<T>` job off of these newly wrapped cancelled entities.

Whether this is through exposing the data stream directly from `AbstractTaskDriver` or continuing to use the specialized "ConfigureCancel" methods is up to you but they should probably just be a convenience wrapper for what is just a basic data stream like any other.

jkeon commented:
Updated for you to try. 

You can now call `taskDriver.ConfigureDriverJobWhenCancelComplete` and the job type is `ITaskJobForDefer<CancelComplete>`

###Tag Alongs

- Fixed a pretty serious access bug in acquiring readers and writers outside the Task System

mbaker3 reviewed (approved):
Tested, works, approved
